### ISSUE TITLE
Memory leak if EnableMetrics = true

### ISSUE THREAD

**Issue by oceandrama** [graphql-dotnet/graphql-dotnet#1324]
Yesterday I noticed a large memory leak of my web-applcation. I dumped it through dotMemory and found a lot of System.Strings. In the dominators tab I found this:
![image](https://user-images.githubusercontent.com/23439079/66200840-97f8bb00-e6aa-11e9-8f12-0b74707c8c9a.png)
PerfRecord is using only in Metrics class.
After disabling metrics memory leak stopped.

**Comment by sungam3r** [comment]
Which versions of graphql packages stack do you use?

**Comment by sungam3r** [comment]
Try the latest preview version - https://www.nuget.org/packages/GraphQL/3.0.0-preview-1271

**Comment by oceandrama** [comment]
I'm using GraphQL.Server.Core 3.5.0-alpha0027. It depends on GraphQL (>= 3.0.0-preview-1268). If I install GraphQL 3.0.0-preview-1271 will it be used as dependency of GraphQL.Server.Core?

**Comment by sungam3r** [comment]
3.0.0-preview-1268 is ok. I see `Metrics` class and its field `private readonly ConcurrentBag<PerfRecord> _records;` - it is on top of your picture. Can you show object graph upper levels?

**Comment by sungam3r** [comment]
![изображение](https://user-images.githubusercontent.com/21261007/66254863-16c22680-e785-11e9-8e0c-d30c7242b972.png)
Of interest is the unfolding up the chain of these objects.

**Comment by sungam3r** [comment]
@oceandrama Please try again with the new package.

**Comment by rolfwessels** [comment]
Hi, 

I'm still having some memory issues (or my expectation could just be wrong). In our ECS cluster I can see memory slowly increase on every graphql call.

![image](https://user-images.githubusercontent.com/643179/70805685-36e41680-1dc2-11ea-99ce-22acb3431945.png)
The gray and the yellow are both graphql-dotnet services. The other are just plain webapi services.
(after a quick bounce they just keep climbing again)

I can reproduce it by just downloading the https://github.com/graphql-dotnet/examples/tree/master/src/AspNetCore project. Updating the package to 
` <PackageReference Include="GraphQL" Version="3.0.0-preview-1352" />`
and setting
`_.EnableMetrics = false;`

The new playground keeps polling the schema and I can see the memory climb on every call.

![image](https://user-images.githubusercontent.com/643179/70803911-ccc97280-1dbd-11ea-96e6-8af2c943b75f.png)

Once the first GC happens the memory seems to plateau out but still climbs slowly (after 15 minutes it was around 210mb. I made no other calls to graphql, it was purely the playground polling .

Expectation:
* When metrics is not enabled I would expect an initial memory increase because of the initial schema loading, I would not expect memory to increases on subsequent calls. 
* I feel that 210mb is quite high for such simple and clean example app. (happy to be wrong about this one)

Hope that helps.

(Was not sure if I need to open new issue or it could be related to this issue.)

**Comment by sungam3r** [comment]
Your first picture indicates that the allocated memory was eventually freed. Growth may be due to heuristics followed by the garbage collector. The behavior of the garbage collector varies on client and server operating systems and depends on many conditions. An increase in memory consumption alone does not indicate leaks. There is a gradual increase in the volume of the managed heap, while the collector allows it. Then there is a one-time garbage collection. This is faster than if the assembly took place more often.

> it was purely the playground polling

Do you imagine the amount of data that is generated as a result of such queries? Depending on the size of the schema, this may be megabytes per request.

In any case, in order to help you, it is impossible to do without creating a memory dump by any profiler. Only after that it will be possible to objectively understand the objects that are somehow held by the application.
